upup/pkg/fi/assettasks/copyfile.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assettasks
+
+import (
+	"fmt"
+
+	"net/url"
+
+	"bytes"
+
+	"github.com/golang/glog"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/util/pkg/hashing"
+	"k8s.io/kops/util/pkg/vfs"
+	"strings"
+)
+
+// CopyFile copies an from a source file repository, to a target repository,
+// typically used for highly secure clusters.
+//go:generate fitask -type=CopyFile
+type CopyFile struct {
+	Name       *string
+	SourceFile *string
+	TargetFile *string
+	// @justinsb not sure I like this but the problem is knowing if the sha is a file or a string
+	// so I did a location or a sha.  We have both SHAs in remote file locations, and SHAs in strings
```

Comment:
Some of our files do not have an SHA file, the code in urls.go for instance, handles that.  Some times the SHA URL for a file, and sometimes the SHA is a string.